test/integration/master/kube_apiserver_test.go

```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
 	}
 }
 
+type delayedCheck struct {
+	healthLock sync.Mutex
+	isHealthy  bool
+}
+
+func (h *delayedCheck) Name() string {
+	return "delayed-check"
+}
+
+func (h *delayedCheck) Check(req *http.Request) error {
+	h.healthLock.Lock()
+	defer h.healthLock.Unlock()
+	if h.isHealthy {
+		return nil
+	}
+	return fmt.Errorf("isn't healthy")
+}
+
+func (h *delayedCheck) makeHealthy() {
+	h.healthLock.Lock()
+	defer h.healthLock.Unlock()
+	h.isHealthy = true
+}
+
+func TestDelayedHealthzAndReadyz(t *testing.T) {
+	hc := &delayedCheck{}
+	instanceOptions := &kubeapiservertesting.TestServerInstanceOptions{
+		InjectedHealthzChecker: hc,
+	}
+	server := kubeapiservertesting.StartTestServerOrDie(t, instanceOptions, []string{"--post-start-hook-grace-period", "5s"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+
+	client, err := kubernetes.NewForConfig(server.ClientConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	readyzRes := client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do()
+	readyzStatus := 0
+	readyzRes.StatusCode(&readyzStatus)
+	if readyzStatus == 200 {
+		t.Fatalf("Readyz should start unhealthy")
+	}
```

Comment:
how about a helper to condense those repeated 7 lines to one?

Reply:
so funny, i actually had this last night but it was lost in one of my changes.